ce/util/evaluator.ts

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { delimiter } from 'path';
+import { createSandbox } from '../util/safeEval';
+import { isPrimitive } from './checks';
+
+/** sandboxed eval function for evaluating expressions */
+const safeEval: <T>(code: string, context?: any) => T = createSandbox();
+
+function proxifyObject(obj: Record<string, any>): any {
+  return new Proxy(obj, {
+    get(target, prop) {
+      if (typeof prop === 'string') {
+        let result = target[prop];
+        // check for a direct match first
+        if (!result) {
+          // go thru the properties and check for a case-insensitive match
+          for (const each of Object.keys(target)) {
+            if (each.toLowerCase() === prop.toLowerCase()) {
+              result = target[each];
+              break;
+            }
+          }
+        }
+        if (result) {
+          if (Array.isArray(result)) {
+            return result;
+          }
+          if (typeof result === 'object') {
+            return proxifyObject(result);
+          }
+          if (isPrimitive(result)) {
+            return result;
+          }
+        }
+        return undefined;
+      }
+    },
+  });
+
+}
+
+export class Evaluator {
+  private activation: any;
+  private host: any;
+
+  constructor(private artifact: Record<string, string>, host: Record<string, any>, activation: Record<string, any>) {
+    this.host = proxifyObject(host);
+    this.activation = proxifyObject(activation);
+
+  }
+
+  evaluate(text: string | undefined): string | undefined {
+    if (!text || text.indexOf('$') === -1) {
+      // quick exit if no expression or no variables
+      return text;
+    }
+
+    // $$ -> escape for $
+    text = text.replace(/\$\$/g, '\uffff');
+
+    // $0 ... $9 -> replace contents with the values from the artifact
+    text = text.replace(/\$([0-9])/g, (match, index) => this.artifact[match] || match);
```

Comment:
Huh? What is `artifact[0]` supposed to be?

Reply:
changed to `artifactData` (it's data from the artifact)